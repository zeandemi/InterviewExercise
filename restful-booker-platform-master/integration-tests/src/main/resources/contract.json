{
  "hotelid": 1,
  "firstname": "James",
  "lastname": "Dean",
  "totalprice": 100,
  "depositpaid": true,
  "bookingdates": {
    "checkin": "2018-02-26",
    "checkout": "2018-02-26"
  }
}