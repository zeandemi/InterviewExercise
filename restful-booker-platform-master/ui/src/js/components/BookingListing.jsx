import React from 'react';
import DatePicker from 'react-datepicker';
import moment from 'moment';

export default class BookingListing extends React.Component {

    constructor(){
        super();

        this.state = {
            allowEdit : false,
            booking : {}
        }

        this.doDelete = this.doDelete.bind(this);
        this.enableEdit = this.enableEdit.bind(this);
        this.disableEdit = this.disableEdit.bind(this);
        this.handleStartChange = this.handleStartChange.bind(this);
        this.handleEndChange = this.handleEndChange.bind(this);
        this.doEdit = this.doEdit.bind(this);
    }

    componentDidMount(){
        this.setState({
            booking : this.props.booking
        })
    }

    doDelete(){
        fetch('http://' + window.location.hostname + ':3000/booking/' + this.props.booking.bookingid, {
			method: 'DELETE',
			credentials: 'include',
        })
        .then(res => {
            if(res.status == 202){
                this.props.fetchHotelDetails();
            }
        })
        .catch(e => console.log(e))
    }

    enableEdit(){
        this.setState({ allowEdit : true });
    }

    disableEdit(){
        this.setState({ allowEdit : false });
    }

    doEdit(){
        fetch('http://' + window.location.hostname + ':3000/booking/' + this.props.booking.bookingid, {
			method: 'PUT',
			headers: {
				'Accept': 'application/json',
				'Content-Type': 'application/json'
			},
            credentials: 'include',
            body : JSON.stringify(this.state.booking)
        })
        .then(res => res.json())
        .then(res => {
            this.setState({allowEdit : false});
            this.props.fetchHotelDetails();
        })
        .catch(e => console.log(e));
    }

    handleStartChange(date) {
        this.setState({
            booking : {
                bookingdates: {
                    checkin : date,
                    checkout : this.state.booking.bookingdates.checkout
                }
            }
        });
    }

    handleEndChange(date) {
        this.setState({
            booking : {
                bookingdates: {
                    checkin : this.state.booking.bookingdates.checkin,
                    checkout : date
                }
            }
        });
    }

    render(){
        let buttons = null;
        let booking = null;

        if(this.props.isAuthenticated){
            buttons = <div className="col-sm-1">
                        <span className="glyphicon glyphicon-pencil bookingEdit" onClick={this.enableEdit} style={{paddingRight: 10 + "px"}}></span>
                        <span className="glyphicon glyphicon-trash bookingDelete" onClick={this.doDelete}></span>
                      </div>
        } else {
            buttons = <div className="col-sm-1"></div>
        }

        if(this.state.allowEdit){

            booking = <div>
                        <div className="col-sm-2"><input type="text" defaultValue={this.props.booking.firstname} onChange={val => this.state.booking.firstname = val.target.value} /></div>
                        <div className="col-sm-2"><input type="text" defaultValue={this.props.booking.lastname} onChange={val => this.state.booking.lastname = val.target.value} /></div>
                        <div className="col-sm-1"><input type="text" defaultValue={this.props.booking.totalprice} onChange={val => this.state.booking.totalprice = val.target.value} /></div>
                        <div className="col-sm-2">
                            <select defaultValue={this.props.booking.depositpaid} onChange={val => this.state.depositpaid = val.target.value}>
                                <option value="false">false</option>
                                <option value="true">true</option>
                            </select>
                        </div>
                        <div className="col-sm-2"><DatePicker selected={moment(this.state.booking.bookingdates.checkin)} onChange={this.handleStartChange} dateFormat="YYYY-MM-DD" /></div>
                        <div className="col-sm-2"><DatePicker selected={moment(this.state.booking.bookingdates.checkout)} onChange={this.handleEndChange} dateFormat="YYYY-MM-DD" /></div>
                        <div className="col-sm-1">
                            <span className="glyphicon glyphicon-ok confirmBookingEdit" onClick={this.doEdit} style={{paddingRight : 10 + "px"}}></span>
                            <span className="glyphicon glyphicon-remove exitBookingEdit" onClick={this.disableEdit}></span>
                        </div>
                    </div>
        } else {
            booking = <div>
                        <div className="col-sm-2"><p>{this.props.booking.firstname}</p></div>
                        <div className="col-sm-2"><p>{this.props.booking.lastname}</p></div>
                        <div className="col-sm-1"><p>{this.props.booking.totalprice}</p></div>
                        <div className="col-sm-2"><p>{String(this.props.booking.depositpaid)}</p></div>
                        <div className="col-sm-2"><p>{this.props.booking.bookingdates.checkin.split('T')[0]}</p></div>
                        <div className="col-sm-2"><p>{this.props.booking.bookingdates.checkout.split('T')[0]}</p></div>
                        {buttons}
                      </div>
        }

        return(
            <div className={"row detail booking-" + this.props.booking.hotelid}>
                {booking}
            </div>
        )
    }

}