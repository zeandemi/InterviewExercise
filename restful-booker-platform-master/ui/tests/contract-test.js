import React from 'react';
import BookingListing from '../src/js/components/BookingListing.jsx';
import contract from '../../integration-tests/src/main/resources/contract.json';

// We first declare our test by using Jests test function and provide it
// with a name and then an anonymous function which will run our test
test('Hotels list component', () => {

    // First we create our component that we want to check. By using
    // enzymes shallow function and passing it the React component
    // we want to create, Enzyme will create a headless DOM and render
    // the React component inside. 
    //
    // You'll notice we are passing parameters to the component namely 
    // the contract we imported at the top of test. If the contract 
    // has been changed to make a different check pass we expect the 
    // check to fail and show what's changed.
    //
    // Also we pass isAuthenticated to put it into a mode where it 
    // renders as if we were logged in.
    const booking = shallow(
        <BookingListing booking={contract} isAuthenticated={true}/>
    );

    // We then compare what the component has done to a previously saved
    // state that lives in the __snapshots__ folder. If anything has changed
    // it will raise an alert.
    expect(booking).toMatchSnapshot();
});