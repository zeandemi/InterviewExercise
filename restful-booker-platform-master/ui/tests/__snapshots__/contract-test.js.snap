// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hotels list component 1`] = `
<div
  className="row detail booking-1"
>
  <div>
    <div
      className="col-sm-2"
    >
      <p>
        James
      </p>
    </div>
    <div
      className="col-sm-2"
    >
      <p>
        Dean
      </p>
    </div>
    <div
      className="col-sm-1"
    >
      <p>
        100
      </p>
    </div>
    <div
      className="col-sm-2"
    >
      <p>
        true
      </p>
    </div>
    <div
      className="col-sm-2"
    >
      <p>
        2018-02-26
      </p>
    </div>
    <div
      className="col-sm-2"
    >
      <p>
        2018-02-26
      </p>
    </div>
    <div
      className="col-sm-1"
    >
      <span
        className="glyphicon glyphicon-pencil bookingEdit"
        onClick={[Function]}
        style={
          Object {
            "paddingRight": "10px",
          }
        }
      />
      <span
        className="glyphicon glyphicon-trash bookingDelete"
        onClick={[Function]}
      />
    </div>
  </div>
</div>
`;
